packages/react-databrowser/src/components/databrowser/hooks/use-keys.tsx

```diff
@@ -0,0 +1,83 @@
+import { createContext, useCallback, useContext, useMemo, type PropsWithChildren } from "react"
+import { useDatabrowserStore } from "@/store"
+import { useInfiniteQuery, type UseInfiniteQueryResult } from "@tanstack/react-query"
+
+import { useFetchKeys, type RedisKey } from "./use-fetch-keys"
+
+const KeysContext = createContext<
+  | {
+      keys: RedisKey[]
+      query: UseInfiniteQueryResult
+      refetch: () => void
+    }
+  | undefined
+>(undefined)
+
+export const FETCH_KEYS_QUERY_KEY = "use-fetch-keys"
+
+export const KeysProvider = ({ children }: PropsWithChildren) => {
+  const { search: searchState } = useDatabrowserStore()
+
+  const search = useMemo(
+    () => ({
+      key: searchState.key.includes("*") ? searchState.key : `*${searchState.key}*`,
+      type: searchState.type,
+    }),
+    [searchState]
+  )
+
+  const { getPage, resetCache } = useFetchKeys(search)
+
+  const query = useInfiniteQuery({
+    queryKey: [FETCH_KEYS_QUERY_KEY, search],
+    initialPageParam: 0,
+    queryFn: async ({ pageParam: pageIndex }) => {
+      return getPage(pageIndex)
+    },
+    select: (data) => data,
+    getNextPageParam: (lastPage, __, lastPageIndex) => {
+      return lastPage.hasNextPage ? lastPageIndex + 1 : undefined
+    },
+  })
+
+  const refetch = useCallback(() => {
+    resetCache()
+    query.refetch()
+  }, [query, resetCache])
+
+  const keys = useMemo(() => {
+    const keys = query.data?.pages.flatMap((page) => page.keys) ?? []
+
+    const keysSet = new Set(keys.map(([key, _]) => key))
```

Comment:
can a key exist twice, why do we take the set here?

Reply:
It's because I modify the cache when adding a new key, to make it appear at the top. But if the user scrolls down and sees that key himself, there might be duplicates.
And I am not sure if redis can give duplicates if keys get deleted etc. while scanning but it's a general safeguard